### DISCUSSION TITLE
How to Check Out Source Controlled Files

### DISCUSSION THREAD

**Discussion by AssisiSoft** [VsixCommunity/Community.VisualStudio.Toolkit#217] [Q&A]
I used to do this to check out files from source control.  Always had issues with files not being fully checked out in time for edits.  DoEvents() seemed to fix the issue.  Is there a better way?

Solution			= DTE.Solution;
ProjectItem		= Solution.FindProjectItem(File);
Document		= ProjectItem.Document;
Window			= Document.ActiveWindow;
Window			.Activate();
DTE.ExecuteCommand("EditorContextMenus.CodeWindow.CheckOutforEdit");
//System.Threading.Thread.Sleep(100);
Application.DoEvents();

**Comment by madskristensen** [comment]
You can do something like this:

```csharp
if (dte.SourceControl.IsItemUnderSCC(fileName) && !dte.SourceControl.IsItemCheckedOut(fileName))
{
     dte.SourceControl.CheckOutItem(fileName);
}
```

**Comment by AssisiSoft** [comment]
Working but what about showing the file after its checked out?
			Doc = await VS.Documents.OpenViaProjectAsync(Path+Item);
 			await Doc.WindowFrame.ShowAsync();

if I open many files in a row, check out fails.  I think because Windows events are not processing.  Is there an equivalent to Application.DoEvents()?

**Comment by AssisiSoft** [comment]
More specifically, here is my check out routine.  CheckOutItem() is successful but IsItemCheckedOut() returns false right after.  Is there some Windows events that need to process because DTE is a COM object?  How do I let them process between the two calls?

    	private async Task<bool>			CheckOutAsync						(string Item)
    	{
			await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
			if (!DTE.SourceControl.IsItemUnderSCC(Path+Item))														throw new Exception(Item+" Not Under SCC");
			if (DTE.SourceControl.IsItemCheckedOut(Path+Item))					return true;
			DTE.SourceControl.CheckOutItem(Path+Item);
			if (!DTE.SourceControl.IsItemCheckedOut(Path+Item))														throw new Exception("Check out failed on '"+Item);
			return true;
    	}

**Comment by NeVeSpl** [comment]
Is it possible to do checkout without using DTE?

**Comment by madskristensen** [comment]
I'm not sure what the API is for that. Perhaps @taysser-gherfal knows?
